node-uniffi/src/lib.rs

```diff
@@ -0,0 +1,313 @@
+//! Native library providing Rust to mobile language bindings for the Lumina node.
+//!
+//! This crate uses Mozillas UniFFI to generate Swift and Kotlin bindings for the Lumina node,
+//! allowing it to be used from iOS and Android applications.
+#![cfg(not(target_arch = "wasm32"))]
+
+mod error;
+mod types;
+
+use celestia_types::ExtendedHeader;
+use error::{LuminaError, Result};
+use lumina_node::{
+    blockstore::RedbBlockstore, events::EventSubscriber, node::PeerTrackerInfo, store::RedbStore,
+    Node,
+};
+use std::str::FromStr;
+use tendermint::hash::Hash;
+use tokio::sync::{Mutex, RwLock};
+use types::{NetworkInfo, NodeConfig, NodeEvent, PeerId, SyncingInfo};
+use uniffi::Object;
+
+uniffi::setup_scaffolding!();
+
+lumina_node::uniffi_reexport_scaffolding!();
+
+/// The main Lumina node that manages the connection to the Celestia network.
+#[derive(Object)]
+pub struct LuminaNode {
+    node: RwLock<Option<Node<RedbBlockstore, RedbStore>>>,
+    events_subscriber: Mutex<Option<EventSubscriber>>,
+    config: NodeConfig,
+}
+
+#[uniffi::export(async_runtime = "tokio")]
+impl LuminaNode {
+    /// Sets a new connection to the Lumina node for the specified network.
+    #[uniffi::constructor]
+    pub fn new(config: NodeConfig) -> Result<Self> {
+        Ok(Self {
+            node: RwLock::new(None),
+            events_subscriber: Mutex::new(None),
+            config,
+        })
+    }
+
+    /// Starts the node and connects to the network.
+    pub async fn start(&self) -> Result<bool> {
+        let mut node_lock = self.node.write().await;
+        if node_lock.is_some() {
+            return Err(LuminaError::network("Node is already running"));
+        }
+
+        let builder = self.config.clone().into_node_builder().await?;
+        let (new_node, subscriber) = builder.start_subscribed().await?;
+
+        *self.events_subscriber.lock().await = Some(subscriber);
+        *node_lock = Some(new_node);
+
+        Ok(true)
+    }
+
+    /// Stops the running node and closes all network connections.
+    pub async fn stop(&self) -> Result<()> {
+        let mut node = self.node.write().await;
+        if let Some(node) = node.take() {
+            node.stop().await;
+            Ok(())
+        } else {
+            Err(LuminaError::network("Node is already stopped"))
+        }
+    }
+
+    /// Checks if the node is currently running.
+    pub async fn is_running(&self) -> bool {
+        self.node.read().await.is_some()
+    }
+
+    /// Gets the local peer ID as a string.
+    pub async fn local_peer_id(&self) -> Result<String> {
+        let node = self.node.read().await;
+        let node = node
+            .as_ref()
+            .ok_or(LuminaError::network("Node not initialized"))?;
```

Comment:
```suggestion
            .ok_or(LuminaError::NodeNotRunning)?;
```
this also applies on all other occurences below